[13.0] Add option to limit replenishment to free quantity

guewen opened the pull request:
This new feature targets buffers of type 'distributed'.

* Store the source location where the replenishment of a buffer should
  occur.
* The source location is taken from the default first route, so we
  assume the default route is a replenishment, otherwise the computation
  is skipped.
* When `stock_buffer_route` is installed, the route configured on the
  buffer is used to find the replenishment source location.
* A computed field shows the free quantity in the replenishment source
  location.
* When a buffer has no quantity in the replenishment source location, a
  new warning is displayed in the tree view.
* A new option on the buffer profile limits the quantity that can be
  replenished in the procurement wizard to the actual free quantity in
  the replenishment source location. The quantity we can procure for
  this profile will be min(recommended quantity, free quantity).

This is a huge win to avoid backorders for transfers created for
replenishments: we create transfers only for the quantity we have, and
if we do not have this quantity in practice, it reveals a stock issue.

Commit:
Add option to limit replenishment to free quantity

This new feature targets buffers of type 'distributed'.

* Store the source location where the replenishment of a buffer should
  occur.
* The source location is taken from the default first route, so we
  assume the default route is a replenishment, otherwise the computation
  is skipped.
* When `stock_buffer_route` is installed, the route configured on the
  buffer is used to find the replenishment source location.
* A computed field shows the free quantity in the replenishment source
  location.
* When a buffer has no quantity in the replenishment source location, a
  new warning is displayed in the tree view.
* A new option on the buffer profile limits the quantity that can be
  replenished in the procurement wizard to the actual free quantity in
  the replenishment source location. The quantity we can procure for
  this profile will be min(recommended quantity, free quantity).

This is a huge win to avoid backorders for transfers created for
replenishments: we create transfers only for the quantity we have, and
if we do not have this quantity in practice, it reveals a stock issue.

Commit:
Add index on ddmrp_history.buffer_id

There is a frequent search on it

jbaudoux reviewed (changes requested):
Fails with "Quantity must be positive." when the buffer is configured in auto-procure but the free qty in source is 0 as it prevents to create a procurement of quantity 0. This error prevent also the refresh of the buffer (as the refresh triggers the auto-procure that fails).

Commit:
Fix error in auto_procure when < free qty

When using the option to limit the procurement to free quantity, and the
free quantity is zero, the auto procurement tries to create procurement
for zero and it raises an error.

guewen commented:
@jbaudoux fixed in my last commit

simahawk reviewed (approved):
LG

LoisRForgeFlow reviewed (approved):
Tested, works smoothly :+1:

LoisRForgeFlow commented:
/ocabot merge minor
